core/services/ocr2/delegate.go

```diff
@@ -650,34 +656,87 @@ func (d *Delegate) ServicesForSpec(jb job.Job) ([]job.ServiceCtx, error) {
+		}
+
+		if cfg.MaxServiceWorkers != 0 {
+			conf.MaxServiceWorkers = cfg.MaxServiceWorkers
+		}
+
+		if cfg.ServiceQueueLength != 0 {
+			conf.ServiceQueueLength = cfg.ServiceQueueLength
+		}
+
+		exec, _ := executer.NewExecuter(
+			log.New(w, "[automation-plugin-executer] ", log.Lshortfile),
+			rgstry,
+			encoder,
+			conf.MaxServiceWorkers,
+			conf.ServiceQueueLength,
+			conf.CacheExpiration,
+			conf.CacheEvictionInterval,
+		)
+		if err2 != nil {
+			return nil, errors.Wrap(err2, "failed to create automation pipeline executer")
+		}
+
+		condObs := &polling.PollingObserverFactory{
+			Logger:   log.New(w, "[automation-plugin-conditional-observer] ", log.Lshortfile),
+			Source:   rgstry,
+			Heads:    rgstry,
+			Executer: exec,
+			Encoder:  encoder,
+		}
+
+		coord := &coordinator.CoordinatorFactory{
+			Logger:     log.New(w, "[automation-plugin-coordinator] ", log.Lshortfile),
+			Encoder:    encoder,
+			Logs:       logProvider,
+			CacheClean: conf.CacheEvictionInterval,
+		}
 
-		conf := ocr2keepers.DelegateConfig{
+		dConf := ocr2keepers.DelegateConfig{
 			BinaryNetworkEndpointFactory: peerWrapper.Peer2,
 			V2Bootstrappers:              bootstrapPeers,
 			ContractTransmitter:          keeperProvider.ContractTransmitter(),
 			ContractConfigTracker:        keeperProvider.ContractConfigTracker(),
 			KeepersDatabase:              ocrDB,
 			LocalConfig:                  lc,
 			Logger:                       ocrLogger,
-			PrefixedLogger:               prefixedLogger,
 			MonitoringEndpoint:           d.monitoringEndpointGen.GenMonitoringEndpoint(spec.ContractID, synchronization.OCR2Automation),
 			OffchainConfigDigester:       keeperProvider.OffchainConfigDigester(),
 			OffchainKeyring:              kb,
 			OnchainKeyring:               kb,
-			HeadSubscriber:               rgstry,
-			Registry:                     rgstry,
-			ReportEncoder:                encoder,
-			PerformLogProvider:           logProvider,
-			CacheExpiration:              cfg.CacheExpiration.Value(),
-			CacheEvictionInterval:        cfg.CacheEvictionInterval.Value(),
-			MaxServiceWorkers:            cfg.MaxServiceWorkers,
-			ServiceQueueLength:           cfg.ServiceQueueLength,
-		}
-		pluginService, err2 := ocr2keepers.NewDelegate(conf)
+			ConditionalObserverFactory:   condObs,
+			CoordinatorFactory:           coord,
+			Encoder:                      encoder,
+			Executer:                     exec,
+			// the following values are not needed in the delegate config anymore
+			CacheExpiration:       cfg.CacheExpiration.Value(),
+			CacheEvictionInterval: cfg.CacheEvictionInterval.Value(),
+			MaxServiceWorkers:     cfg.MaxServiceWorkers,
+			ServiceQueueLength:    cfg.ServiceQueueLength,
+		}
+
+		pluginService, err2 := ocr2keepers.NewDelegate(dConf)
 		if err2 != nil {
-			return nil, errors.Wrap(err, "could not create new keepers ocr2 delegate")
+			return []job.ServiceCtx{}, errors.Wrap(err, "could not create new keepers ocr2 delegate")
```

Comment:
Why change this to return an empty slice when nil results in the same behaviour?

Reply:
Ugh, weird!